File: LeetCode_Algorithm/2095(Delete the Middle Node of a Linked List)/2095.js
function ListNode(val, next) {
  this.val = (val === undefined ? 0 : val)
  this.next = (next == undefined ? null : next)
}

var deleteMiddle = function(head) {
  // 668ms 53% 126.1MB 68%
  if (!head || !head.next) {
    return null
  }
  let slow = head; let fast = head;
  let prev = slow;
  while (fast && fast.next) {
    prev = slow
    slow = slow.next
    fast = fast.next.next
  }
  prev.next = slow.next
  return head
};

File: LeetCode_Algorithm/2095(Delete the Middle Node of a Linked List)/2095.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
from typing import Optional

'''
링크드리스트에서 중간 노드를 삭제하기

# sol1 1666ms 92% 62.9MB 57%
slow와 fast를 만들어서 해결

'''
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return None
        prev = slow = fast = head
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next
        prev.next = slow.next
        return head
